doc/Samplepacket/SAI-Proposal-Samplepacket_truncation.md

```diff
@@ -0,0 +1,206 @@
+
+Currently, SAI does not provide an option for users to configure truncation for sample packets directed to the CPU (`SAI_SAMPLEPACKET_TYPE_SLOW_PATH`).
+
+**Note:** For packets that are sampled to an external collector using `SAI_SAMPLEPACKET_TYPE_MIRROR_SESSION`, truncation can be configured using `SAI_MIRROR_SESSION_ATTR_TRUNCATE_SIZE`. However, this is not applicable for `SAI_SAMPLEPACKET_TYPE_SLOW_PATH`.
+
+---
+
+## SAI API model for sFlow to local collector
+
+
+### API model
+
+The below explains the API sequence for configuring `SAI_SAMPLEPACKET_TYPE_SLOW_PATH`.
+
+
+#### Create Hostif Trap
+
+```c
+sai_attribute_t attr_list[4];
+
+attr_list[0].id = SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE;
+attr_list[0].value.s32 = SAI_HOSTIF_TRAP_TYPE_SAMPLEPACKET;
+
+attr_list[1].id = SAI_HOSTIF_TRAP_ATTR_TRAP_GROUP;
+attr_list[1].value.oid = <trap_group_oid>;
+
+attr_list[2].id = SAI_HOSTIF_TRAP_ATTR_PACKET_ACTION;
+attr_list[2].value.s32 = SAI_PACKET_ACTION_TRAP;
+
+attr_list[3].id = SAI_HOSTIF_TRAP_ATTR_TRAP_PRIORITY;
+attr_list[3].value.u32 = 1;
+
+sai_object_id_t hostif_trap_id;
+sai_create_hostif_trap(&hostif_trap_id, switch_id, 4, attr_list);
+```
+
+#### Create Sample Packet
+
+```c
+sai_attribute_t attr[2];
+
+attr[i].id = SAI_SAMPLEPACKET_ATTR_TYPE;
+attr[i++].value.s32 = SAI_SAMPLEPACKET_TYPE_SLOW_PATH;
+
+attr[i].id = SAI_SAMPLEPACKET_ATTR_SAMPLE_RATE;
+attr[i++].value.u32 = 100000;
+
+
+sai_object_id_t samplepacket_id;
+sai_create_samplepacket(&samplepacket_id, switch_id, i, &attr);
+```
+
+#### Enable Sample Packet on Port
+
+```c
+sai_attribute_t attr;
+
+attr.id = SAI_PORT_ATTR_INGRESS_SAMPLEPACKET_ENABLE;
+attr.value.oid = samplepacket_id;
+
+sai_set_port_attribute(port_oid, &attr);
+```
+
+
+
+
+### Data flow
+
+The below diagram explains the sample packet data flow using SONiC as an example.
+
+```mermaid
+sequenceDiagram
+ASIC ->> Driver : sFlow(sample)
+Driver ->> GENETLINK : encapsulated sample: IIF-INDEX, SAMPLE-RATE etc.
+GENETLINK ->> hsflowd : genetlink multicast message
+hsflowd ->> sFlow_Collector : sFlow datagram
+```
+
+
+## SAI Enhancements to support truncation
+
+
+
+
+```c
+@@ -114,6 +114,26 @@ typedef enum _sai_samplepacket_attr_t
+      */
+     SAI_SAMPLEPACKET_ATTR_MODE,
+
++    /**
++     * @brief Enable truncation of sampled packets
++     *
++     * @type bool
++     * @flags CREATE_AND_SET
++     * @default false
++     */
++    SAI_SAMPLEPACKET_ATTR_TRUNCATE_ENABLE,
++
++    /**
++     * @brief Truncate size. Truncate sampled packets to this size to reduce traffic bandwidth
++     *
++     * This value takes effect when SAI_SAMPLEPACKET_ATTR_TRUNCATE_ENABLE is True.
++     *
++     * @type sai_uint32_t
++     * @flags CREATE_AND_SET
++     * @default internal
++     */
++    SAI_SAMPLEPACKET_ATTR_TRUNCATE_SIZE,
```

Comment:
Qn from @JaiOCP during review on Aug-15-2024:  In SAI\_SAMPLEPACKET\_TYPE\_MIRROR\_SESSION, define the behavior when both SAI\_SAMPLEPACKET\_ATTR\_TRUNCATE\_SIZE and SAI\_MIRROR\_SESSION\_ATTR\_TRUNCATE\_SIZE are given. 

The explanation was to follow #1082: Use SAI\_SAMPLEPACKET\_ATTR\_TRUNCATE\_SIZE if we want to have the truncation specific to a port, and use SAI\_MIRROR\_SESSION\_ATTR\_TRUNCATE\_SIZE if we want the same truncation for all the ports using the same mirror session object.

Jai wanted to do more investigations and get back if the above behavior is agreeable.